Feature/sync team to firebase

Rup-Narayan-Rajbanshi opened the pull request:
## Changes
- Sync team create update data to firebase.


## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [x] tests
- [x] permission checks (tests here too)
- [x] translations

Rup-Narayan-Rajbanshi committed:
chore(firebase): add base abstract class for firebase resources

Rup-Narayan-Rajbanshi committed:
chore(firebase): sync contributor team to firebase

tnagorra commented on apps/common/models.py:
```diff
@@ -2,13 +2,22 @@
 from utils.common import validate_ulid
 
 
+class FirebasePushStatusEnum(models.IntegerChoices):
+    PENDING = 1, "Pending"
+    PROCESSING = 2, "Processing"
+    SUCCESS = 3, "Success"
+    FAILED = 4, "Failed"
```
Not sure why these changes are visible here. Isn't this already implemented in `project/sync-to-firebase`

tnagorra replied:
The history does not look correct.

<img width="245" height="283" alt="Image" src="https://github.com/user-attachments/assets/3dff1440-e653-48de-86e8-e747b90cbbac" />

tnagorra commented on main/config.py:
```diff
@@ -42,6 +42,10 @@ def project_groups(project_id: str | int):
 
+        @staticmethod
+        def contributor_team(contributor_team_id: str | int):
+            return f"/v2/contributor_team/{contributor_team_id}/"
```
```suggestion
            return f"/v2/teams/{contributor_team_id}/"
```

tnagorra commented on apps/contributor/firebase.py:
```diff
@@ -0,0 +1,94 @@
+    contributor_team_data = firebase_models.FbTeam(
+        teamName=contributor_team.name,
+        isArchived=contributor_team.is_archived,
+        teamToken=datetime.now(),  # FIX: What is this used for?
```
Let's make this optional and the type can be str.
Also rename "FIX" to "FIXME(your-username)"

tnagorra commented on apps/contributor/firebase.py:
```diff
@@ -0,0 +1,94 @@
+    contributor_team_ref.set(
+        value={
+            **firebase_utils.serialize(contributor_team_data),
+        },
```
You do not need to destructure dictionary here.

tnagorra commented on apps/contributor/firebase.py:
```diff
@@ -0,0 +1,94 @@
+            firebase_models.FbTeam(
+                teamName=contributor_team.name,
+                isArchived=contributor_team.is_archived,
+                teamToken=datetime.now(),  # FIX: What is this used for?
```
Let's make this optional and the type can be str.
Also rename "FIX" to "FIXME(your-username)"

tnagorra commented on apps/contributor/firebase.py:
```diff
@@ -0,0 +1,94 @@
+
+    try:
+        contributor_team_ref = Config.FIREBASE_HELPER.ref(
+            Config.FirebaseKeys.contributor_team(contributor_team.id),
```
We cannot use "id"
We will need to use "old_id" if it exists.
@thenav56 Should we just fallback to only using "old_id" or "legacy_id"?

tnagorra commented on apps/contributor/firebase.py:
```diff
@@ -0,0 +1,94 @@
+
+
+@shared_task
+def push_contributor_team_to_firebase(contributor_team_id: int):
```
@thenav56 Should we reuse the whole "push_to_firebase" logic for single object?

Rup-Narayan-Rajbanshi commented on apps/common/models.py:
```diff
@@ -63,3 +72,30 @@ class ArchivableResource(Model):
     class Meta(TypedModelMeta):  # type: ignore[reportIncompatibleVariableOverride]
         abstract = True
         ordering = ["-id"]
+
+
+class FirebaseResource(Model):
+    firebase_push_status: int | None = IntegerChoicesField(  # type: ignore[reportAssignmentType]
+        choices_enum=FirebasePushStatusEnum,
+        null=True,
+        blank=True,
+    )
+    firebase_last_pushed = models.DateTimeField(
+        null=True,
+        blank=True,
+        help_text=gettext_lazy("The latest time when project was pushed to firebase"),
+    )
+
+    @property
+    def firebase_push_status_enum(self):
+        if self.firebase_push_status:
+            return FirebasePushStatusEnum(self.firebase_push_status)
+        return None
+
+    def update_firebase_push_status(self, firebase_push_status: FirebasePushStatusEnum, *, commit: bool = True):
+        self.firebase_push_status = firebase_push_status
+        if commit:
+            self.save(update_fields=("firebase_push_status",))
+
+    class Meta(TypedModelMeta):  # type: ignore[reportIncompatibleVariableOverride]
+        abstract = True
```
Let's use this in project and remove the usage in the project before we merge this

Rup-Narayan-Rajbanshi commented on apps/common/models.py:
```diff
@@ -63,3 +72,30 @@ class ArchivableResource(Model):
+    def update_firebase_push_status(self, firebase_push_status: FirebasePushStatusEnum, *, commit: bool = True):
+        self.firebase_push_status = firebase_push_status
+        if commit:
+            self.save(update_fields=("firebase_push_status",))
```
shouldn't we also handle firebase_last_pushed here?

Rup-Narayan-Rajbanshi committed:
chore(firebase): add common func for push_to_firebase

  - Add field old id in teams
  - Refactor project to accept base firebase resources.

Rup-Narayan-Rajbanshi committed:
chore(README): update python version in pre-commit-config

tnagorra commented on apps/contributor/models.py:
```diff
@@ -80,8 +81,9 @@ def __str__(self):
-class ContributorTeam(ArchivableResource, UserResource):  # type: ignore[reportIncompatibleVariableOverride]
+class ContributorTeam(ArchivableResource, UserResource, FirebaseResource):  # type: ignore[reportIncompatibleVariableOverride]
     name = models.CharField(max_length=255)
+    uuid = models.UUIDField(default=uuid.uuid4)
```
1. Let's rename this to "token"
2. Let's make this unique

tnagorra commented on apps/common/models.py:
```diff
@@ -101,3 +111,39 @@ class IconEnum(models.IntegerChoices):
+        firebase_push_status: FirebasePushStatusEnum,
+        firebase_last_pushed: datetime | None = None,
+    ):
+        self.firebase_push_status = firebase_push_status
+        if firebase_push_status == FirebasePushStatusEnum.SUCCESS:
+            self.firebase_last_pushed = firebase_last_pushed
+            self.save(update_fields=("firebase_push_status", "firebase_last_pushed"))
+            return
+        self.save(update_fields=("firebase_push_status",))
```
Let's not remove commit flag

tnagorra replied:
We don'tneed firebase_last_pushed as an argument. Let's just use timezone.now()

tnagorra commented on apps/common/tasks.py:
```diff
@@ -1,17 +1,82 @@
+
+        else:
+            if fb_model is None:
+                logger.error(
+                    "push_to_firebase found did not find %s in firebase when updating a %s",
+                    model_obj._meta.label,
+                    model_obj._meta.label,
+                    extra=log_extra({"model_obj_id": model_obj.pk}),
+                )
```
Let's not use "%s" twice.

tnagorra commented on apps/common/tasks.py:
```diff
@@ -1,17 +1,82 @@
+        else:
+            if fb_model is None:
+                logger.error(
+                    "push_to_firebase found did not find %s in firebase when updating a %s",
```
```suggestion
                    "push_to_firebase did not find %s in firebase when updating a %s",
```

tnagorra reviewed (changes requested):
@thenav56 was right! Got issue with the migration script regarding uniqueness
<img width="1128" height="60" alt="image" src="https://github.com/user-attachments/assets/dfafbd83-ca17-4166-8519-0a3677a43cf9" />
